electron/syntheticsGenerator.ts

```diff
@@ -0,0 +1,364 @@
+    }
+    return false;
+  }
+
+  generateStepStart(name: string) {
+    this.insideStep = true;
+    const formatter = new PlaywrightGenerator.JavaScriptFormatter(this.getDefaultFormatterOffset());
+    formatter.add(`step(${quote(name)}, async () => {`);
+    return formatter.format();
+  }
+
+  generateStepEnd() {
+    if (!this.insideStep) {
+      return '';
+    }
+    this.insideStep = false;
+    const formatter = new PlaywrightGenerator.JavaScriptFormatter(this.getDefaultFormatterOffset());
+    formatter.add(`});`);
+    return formatter.format();
+  }
+
+  generateHeader() {
+    const formatter = new PlaywrightGenerator.JavaScriptFormatter(0);
+    formatter.add(`
+      const { journey, step, expect } = require('@elastic/synthetics');
+
+      journey('Recorded journey', async ({ page, context }) => {`);
+    return formatter.format();
+  }
+
+  generateFooter() {
+    return `});`;
+  }
+
+  /**
+   * Generates JavaScript code from a custom set of steps and nested actions.
+   *
+   * This function makes no assumptions about where steps should be created,
+   * and instead follows the step definitions the caller has defined.
+   * @param steps IR to use for code generation
+   * @returns a list of the code strings outputted by the generator
+   */
+  generateFromSteps(steps: Steps): string {
+    const text = [];
+    if (this.isSuite) {
+      text.push(this.generateHeader());
+    }
+    this.varsToHoist = this.findVarsToHoist(steps);
+    text.push(this.generateHoistedVars());
+    for (const step of steps) {
+      if (step.actions.length === 0) throw Error('Cannot process an empty step');
+      const name = step.name ?? step.actions[0].title ?? actionTitle(step.actions[0].action);
+      text.push(this.generateStepStart(name ?? ''));
+
+      for (const action of step.actions) {
+        const actionText = this.generateAction(action);
+        if (actionText.length) text.push(actionText);
+      }
+
+      text.push(this.generateStepEnd());
+    }
+    if (this.isSuite) {
+      text.push(this.generateFooter());
+    }
+
+    return text.filter(s => !!s).join('\n');
+  }
+
+  generateHoistedVars() {
+    const formatter = new PlaywrightGenerator.JavaScriptFormatter(this.getDefaultFormatterOffset());
+    for (const varName of this.varsToHoist) {
+      formatter.add(`let ${varName};`);
+    }
+    return formatter.format();
+  }
+
+  isVarHoisted(varName: string) {
+    return this.varsToHoist.indexOf(varName) >= 0;
+  }
+
+  getDefaultFormatterOffset() {
+    return this.isSuite ? 2 : 0;
+  }
+
+  /**
+   * We need to hoist any page or popup alias that appears in more than one step.
+   * @param steps the step IR to evaluate
+   * @returns an array that contains the names of all variables that need to be hoisted
+   */
+  findVarsToHoist(steps: Steps): string[] {
+    const pageAliasCounts = steps.reduce((acc, step) => {
+      const aliases = new Set<string>();
```

Comment:
This algorithm can be simplified by having this two thoughts
- No need to distinguish between steps, since no of alias is proportional to the number of pages created by the user. 
- Hoist the page alias no matter what even there is only one step.

Having that assumptions, the code would look like this
- Having a single set that maintains the page alias
- Go over the popups and aliases and add it to the Set
- Produce the generated script except for the `page` object which is maintained by Synthetics.

Reply:
Ok, that is simpler, thanks for the suggestion. My original motivation was to avoid hoisting unless it was necessary, this way we'd keep non-glitchy scripts as close to the original format as possible. But this is indeed a simpler approach and if you're ok with it I'm fine with it as well. Thanks for the detailed comment!

Reply:
https://github.com/elastic/synthetics-recorder/pull/211/commits/638d8bd1b00943faa45c9139f3a41a352b072d1e